stix2/properties.py

```diff
@@ -17,82 +23,109 @@ class Property(object):
-          ValueError if the value is not valid.
-          (Default: if `clean` is defined, it will attempt to call `clean` and
-          return the result or pass on a ValueError that `clean` raises. If
-          `clean` is not defined, this will return `value` unmodified).
+        - Return a value that is valid for this property. If `value` is not
+          valid for this property, this will attempt to transform it first. If
+          `value` is not valid and no such transformation is possible, it should
+          raise a ValueError.
     - `def default(self):`
         - provide a default value for this property.
         - `default()` can return the special value `NOW` to use the current
             time. This is useful when several timestamps in the same object need
             to use the same default value, so calling now() for each field--
             likely several microseconds apart-- does not work.
 
-    Subclasses can instead provide lambda functions for `clean`, and `default`
-    as keyword arguments. `validate` should not be provided as a lambda since
-    lambdas cannot raise their own exceptions.
+    Subclasses can instead provide a lambda function for `default as a keyword
+    argument. `clean` should not be provided as a lambda since lambdas cannot
+    raise their own exceptions.
     """
 
-    def _default_validate(self, value):
+    def _default_clean(self, value):
         if value != self._fixed_value:
             raise ValueError("must equal '{0}'.".format(self._fixed_value))
         return value
 
-    def __init__(self, required=False, fixed=None, clean=None, default=None, type=None):
+    def __init__(self, required=False, fixed=None, default=None, type=None):
         self.required = required
         self.type = type
         if fixed:
             self._fixed_value = fixed
-            self.validate = self._default_validate
+            self.clean = self._default_clean
             self.default = lambda: fixed
-        if clean:
-            self.clean = clean
         if default:
             self.default = default
 
     def clean(self, value):
-        raise NotImplementedError
-
-    def validate(self, value):
-        try:
-            value = self.clean(value)
-        except NotImplementedError:
-            pass
         return value
 
+    def __call__(self, value=None):
+        if value is not None:
+            return value
+
 
 class ListProperty(Property):
 
-    def __init__(self, contained, required=False, element_type=None):
+    def __init__(self, contained, **kwargs):
         """
-        contained should be a type whose constructor creates an object from the value
+        Contained should be a function which returns an object from the value.
         """
-        self.contained = contained
-        self.element_type = element_type
-        super(ListProperty, self).__init__(required)
+        if contained == StringProperty:
```

Comment:
In Python, \_\_call\_\_ lets you do things like:

```python
class A():
    ...

a = A() # calls __init__()
a() # calls __call__()
```

This is needed for ListProperty fields. For some, like AttackPattern.kill_chain_phases, ListProperty.contained is set to the KillChainPhase type (and not an instance of KillChainPhase because of KillChainPhase's required properties). But for others, like Sighting.where_sighted_refs, the ListProperty.contained is set to an instance of ReferenceProperty because where_sighted_refs is limited to references to Identity objects and we have to set the ReferenceProperty.type.

clean() could be a static or class method for some Property types, but not all of them. For example, IDProperty.clean() uses self.required_prefix.

You can indeed create lists of lists, but the innermost ListProperty must specify a type. `ListProperty(ListProperty())` will throw a TypeError because the ListProperty constructor requires a parameter, but `ListProperty(ListProperty(StringProperty))` is fine.